src/macro/eval/evalSsl.ml

```diff
@@ -0,0 +1,142 @@
+open EvalHash
+open EvalValue
+open EvalEncode
+open EvalDecode
+open EvalExceptions
+open Mbedtls
+
+let as_cert vthis = match vthis with
+	| VInstance {ikind = IEmbedtlsCertificate i} -> i
+	| _ -> unexpected_value vthis "Certificate"
+
+let as_config vthis = match vthis with
+	| VInstance {ikind = IEmbedtlsConfig i} -> i
+	| _ -> unexpected_value vthis "Config"
+
+let as_socket vthis = match vthis with
+	| VInstance {ikind = ISocket sock} -> sock
+	| _ -> unexpected_value vthis "NativeSocket"
+
+let as_ctr_drbg vthis = match vthis with
+	| VInstance {ikind = IEmbedtlsCtrDrbg i} -> i
+	| _ -> unexpected_value vthis "CtrDrbg"
+
+let as_entropy vthis = match vthis with
+	| VInstance {ikind = IEmbedtlsEntropy i} -> i
+	| _ -> unexpected_value vthis "Entropy"
+
+let as_ssl vthis = match vthis with
+	| VInstance {ikind = IEmbedtlsSsl ctx} -> ctx
+	| _ -> unexpected_value vthis "Ssl"
+
+let init_constructors add =
+	add key_mbedtls_Certificate
+		(fun _ ->
+			let cert = mbedtls_x509_crt_init() in
+			encode_instance key_mbedtls_Certificate ~kind:(IEmbedtlsCertificate cert)
+		);
+	add key_mbedtls_Config
+		(fun _ ->
+			let cfg = mbedtls_ssl_config_init() in
+			encode_instance key_mbedtls_Config ~kind:(IEmbedtlsConfig cfg)
+		);
+	add key_mbedtls_CtrDrbg
+		(fun _ ->
+			let ctr = mbedtls_ctr_drbg_init() in
+			encode_instance key_mbedtls_CtrDrbg ~kind:(IEmbedtlsCtrDrbg ctr)
+		);
+	add key_mbedtls_Entropy
+		(fun _ ->
+			let entropy = mbedtls_entropy_init() in
+			encode_instance key_mbedtls_Entropy ~kind:(IEmbedtlsEntropy entropy)
+		);
+	add key_mbedtls_Ssl
+		(fun _ ->
+			let ssl = mbedtls_ssl_init() in
+			encode_instance key_mbedtls_Ssl ~kind:(IEmbedtlsSsl ssl)
+		)
+
+let init_fields init_fields builtins =
+	init_fields builtins (["sys";"ssl"],"Mbedtls") [
+		"loadDefaults",vfun1 (fun this ->
+			vint (hx_cert_load_defaults (as_cert this));
+		);
+	] [];
+	init_fields builtins (["mbedtls"],"Certificate") [] [
+		"parse_file",vifun1 (fun this path ->
+			vint (mbedtls_x509_crt_parse_file (as_cert this) (decode_string path));
+		);
+		"parse_path",vifun1 (fun this path ->
+			vint (mbedtls_x509_crt_parse_path (as_cert this) (decode_string path));
+		);
+	];
+	init_fields builtins (["mbedtls"],"Config") [] [
+		"authmode",vifun1 (fun this authmode ->
+			mbedtls_ssl_config_authmode (as_config this) (decode_int authmode);
+			vnull;
+		);
+		"caChain",vifun2 (fun this ca_chain ca_crl ->
+			mbedtls_ssl_conf_ca_chain (as_config this) (as_cert ca_chain) None;
+			vnull;
+		);
+		"defaults",vifun3 (fun this endpoint transport preset ->
+			vint (mbedtls_ssl_config_defaults (as_config this) (decode_int endpoint) (decode_int transport) (decode_int preset));
+		);
+		"free",vifun0 (fun this ->
```

Comment:
This is very un-Haxe like, unfortunately. Might want to initialise the abstract on the C side as a custom block with a finaliser. (`caml_alloc_custom`)

Reply:
It's not that unusual, we have `close` functions on various objects and expect the user to call that to free resources.

But yes, if we can abstract this away in the allocating C functions it certainly won't hurt. In that case we should even remove it from the externs to avoid a duplicate freeing.

Reply:
I've rewritten this to use `caml_alloc_custom` now. I wouldn't exactly say that I know what I'm doing, but following the pattern from https://github.com/project-everest/MLCrypto/blob/master/openssl_stub.c seems to do the trick.